encoding/wave/decode.go

```diff
@@ -0,0 +1,217 @@
+package wave
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"reflect"
+)
+
+type (
+	// Decoder is a WAVE decoder
+	Decoder struct {
+		input io.Reader
+	}
+)
+
+// NewDecoder creates a new WAVE decoder
+func NewDecoder(r io.Reader) *Decoder {
+	return &Decoder{
+		input: r,
+	}
+}
+
+// Decode the WAVE into data array and returns the WAVE header
+func (d *Decoder) Decode(data interface{}) (Header, error) {
+	v := reflect.ValueOf(data)
+	if v.Kind() != reflect.Ptr || v.IsNil() {
+		return Header{}, fmt.Errorf("Decode expects a non-nil pointer")
+	}
+	deref := v.Elem()
+	switch deref.Kind() {
+	case reflect.Slice:
+		// get the underlying type of slice elements
+		switch typ := deref.Type().Elem().Kind(); typ {
+		case reflect.Int16, reflect.Float32:
+			return d.decode(v, typ)
+		default:
+			return Header{}, fmt.Errorf("unknown slice type: %v",
+				deref.Type().Elem().Kind())
+		}
+	}
+
+	return Header{}, fmt.Errorf("impossible to decode wave into %s. "+
+		"Expect *[]int16, *[]float32", v.Kind())
+}
+
+func (d *Decoder) decode(v reflect.Value, typ reflect.Kind) (Header, error) {
+	hdr, err := d.parseHeader()
+	if err != nil {
+		return Header{}, err
+	}
+
+	switch typ {
+	case reflect.Int16:
+		if hdr.RiffChunkFmt.AudioFormat != FormatPCM {
+			return hdr, fmt.Errorf("[]int16 requires a PCM audio")
+		}
+		err = d.decodeInt16(v, hdr.DataBlockSize)
+	case reflect.Float32:
+		if hdr.RiffChunkFmt.AudioFormat != FormatIEEEFloat {
+			return hdr, fmt.Errorf("[]float32 requires a IEEEFloat format")
+		}
+		err = d.decodeFloat32(v, hdr.DataBlockSize)
+	}
+	return hdr, err
+}
+
+func (d *Decoder) decodeInt16(v reflect.Value, datasz uint32) error {
+	dataptr := v.Interface().(*[]int16)
+
+	const typesize = 2
+	var bytesRead uint32
+	for bytesRead < datasz {
+		var buf [typesize]byte
+		n, err := d.input.Read(buf[:])
+		if err != nil {
+			return err
+		}
+		if n != typesize {
+			return fmt.Errorf("corrupted audio")
+		}
+
+		sample := int16(binary.LittleEndian.Uint16(buf[:]))
+		*dataptr = append(*dataptr, sample)
+		bytesRead += typesize
+	}
+
+	return nil
+}
+
+func (d *Decoder) decodeFloat32(v reflect.Value, datasz uint32) error {
+	const maxval float32 = 1.0
+	const minval float32 = -1.0
+	const typesz = 4
+
+	dataptr := v.Interface().(*[]float32)
+	var bytesRead uint32
+	for bytesRead < datasz {
```

Comment:
No, because reading till EOF you can read chunks other than data. The "fmt " chunk says the size in bytes of the "data" chunk. Accordingly to the spec, chunks other than "fmt " could arise in any order. We cannot assume data is the last.

Reply:
that's why I'm not handling the EOF case, because if there's an EOF than we have a data corruption.